### PR TITLE
Namespace must estimate and requeue content deletions

### PR DESCRIPTION
**Pull request by smarterclayton**
When using graceful deletion, retry repeatedly until items are completed.
Assumes finalization is managed by the underlying types (if a gracefully
deleted pod is not cleaned up, namespace deletion will not complete).

Built on top of #12404

Extracted from #9165

### CODE REVIEW

**Comment by smarterclayton** [comment]
@lavalamp @nikhiljindal this was already code reviewed once in graceful, but needs signoff to reduce the size of #9165

File: pkg/controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -55,16 +57,36 @@ func NewNamespaceController(kubeClient client.Interface, resyncPeriod time.Durat
 		framework.ResourceEventHandlerFuncs{
 			AddFunc: func(obj interface{}) {
 				namespace := obj.(*api.Namespace)
-				err := syncNamespace(kubeClient, *namespace)
-				if err != nil {
-					glog.Error(err)
+				if err := syncNamespace(kubeClient, *namespace); err != nil {
+					if estimate, ok := err.(*contentRemainingError); ok {
+						go func() {
+							t := estimate.Estimate/2 + 1
```
Comment (nikhiljindal) [inline pkg/controller/namespace/namespace_controller.go]: If the remaining content is going to take 10 secs to delete, then why not wait 10 secs? Why wait just 6 secs?

File: pkg/controller/namespace/namespace_controller.go [hunk 2]
```diff
@@ -41,7 +42,8 @@ type NamespaceController struct {
 
 // NewNamespaceController creates a new NamespaceController
 func NewNamespaceController(kubeClient client.Interface, resyncPeriod time.Duration) *NamespaceController {
-	_, controller := framework.NewInformer(
+	var controller *framework.Controller
```
Comment (nikhiljindal) [inline pkg/controller/namespace/namespace_controller.go]: This should not be required?

**Comment by nikhiljindal** [comment]
Thanks a lot for splitting the code.
Mostly LGTM. Just a comment on why not wait for the estimated time.

File: pkg/controller/namespace/namespace_controller.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/controller/namespace/namespace_controller.go]: The estimate is the maximum, most things will terminate fast and a few will
go longer.  I *believe* the distribution will be binomial with a strong
peak around a few seconds, but our default is 30 seconds for pods.   So I
didn't want to wait 30, and /2 felt like a good compromise.  +1 was to
ensure that we wouldn't wait 0 in scale up cases.  Will add a comment.

On Mon, Aug 10, 2015 at 3:21 AM, Nikhil Jindal <notifications@github.com>
wrote:

> In pkg/controller/namespace/namespace_controller.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12406#discussion_r36606759>
> :
>
> > @@ -55,16 +57,36 @@ func NewNamespaceController(kubeClient client.Interface, resyncPeriod time.Durat
> >  		framework.ResourceEventHandlerFuncs{
> >  			AddFunc: func(obj interface{}) {
> >  				namespace := obj.(*api.Namespace)
> > -				err := syncNamespace(kubeClient, *namespace)
> > -				if err != nil {
> > -					glog.Error(err)
> > +				if err := syncNamespace(kubeClient, *namespace); err != nil {
> > +					if estimate, ok := err.(*contentRemainingError); ok {
> > +						go func() {
> > +							t := estimate.Estimate/2 + 1
>
> If the remaining content is going to take 10 secs to delete, then why not
> wait 10 secs? Why wait just 6 secs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12406/files#r36606759>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

File: pkg/controller/namespace/namespace_controller.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/controller/namespace/namespace_controller.go]: We need controller as a variable to use from the requeue mechanism (line 66)

File: pkg/controller/namespace/namespace_controller.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/controller/namespace/namespace_controller.go]: Comment to that effect added.

On Mon, Aug 10, 2015 at 11:16 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> The estimate is the maximum, most things will terminate fast and a few
> will go longer.  I *believe* the distribution will be binomial with a
> strong peak around a few seconds, but our default is 30 seconds for pods.
> So I didn't want to wait 30, and /2 felt like a good compromise.  +1 was to
> ensure that we wouldn't wait 0 in scale up cases.  Will add a comment.
>
> On Mon, Aug 10, 2015 at 3:21 AM, Nikhil Jindal <notifications@github.com>
> wrote:
>
>> In pkg/controller/namespace/namespace_controller.go
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12406#discussion_r36606759>
>> :
>>
>> > @@ -55,16 +57,36 @@ func NewNamespaceController(kubeClient client.Interface, resyncPeriod time.Durat
>> >  		framework.ResourceEventHandlerFuncs{
>> >  			AddFunc: func(obj interface{}) {
>> >  				namespace := obj.(*api.Namespace)
>> > -				err := syncNamespace(kubeClient, *namespace)
>> > -				if err != nil {
>> > -					glog.Error(err)
>> > +				if err := syncNamespace(kubeClient, *namespace); err != nil {
>> > +					if estimate, ok := err.(*contentRemainingError); ok {
>> > +						go func() {
>> > +							t := estimate.Estimate/2 + 1
>>
>> If the remaining content is going to take 10 secs to delete, then why not
>> wait 10 secs? Why wait just 6 secs?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12406/files#r36606759>
>> .
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by nikhiljindal** [comment]
LGTM, thanks!

**Pull request merged**